pkg/crypto/rings/client.go

```diff
@@ -194,3 +206,52 @@ func (rc *ringClient) addressesToPubKeys(
 	}
 	return pubKeys, nil
 }
+
+// GetRingAddressesAtBlock returns the active gateway delegations for the given
+// application and target block height while accounting for the pending undelegations.
+// The ring addresses slice is reconstructed by adding back the past delegated gateways
+// that have been undelegated after the target session end height.
+func GetRingAddressesAtBlock(app *apptypes.Application, blockHeight int64) []string {
+	// Get the target session end height at which we want to get the active delegations.
+	targetSessionEndHeight := uint64(sessionkeeper.GetSessionEndBlockHeight(blockHeight))
+	// Get the current active delegations for the application and use them as a base.
+	activeDelegationsAtHeight := app.DelegateeGatewayAddresses
+
+	// Use a map to keep track of the gateways addresses that have been added to
+	// the active delegations slice to avoid duplicates.
+	addedDelegations := make(map[string]bool)
+
+	// Iterate over the pending undelegations recorded at their respective block
+	// height and check whether to add them back as active delegations.
+	for pendingUndelegationHeight, undelegatedGateways := range app.PendingUndelegations {
+		// If the pending undelegation happened BEFORE the target session end height,
+		// skip it, as it became effective before the target session end height.
+		if targetSessionEndHeight > pendingUndelegationHeight {
+			continue
+		}
+		// Add back any gateway address  that was undelegated after the target session
+		// end height, as we consider it not happening yet relative to the target height.
+		for _, gatewayAddress := range undelegatedGateways.GatewayAddresses {
+			if _, ok := addedDelegations[gatewayAddress]; ok {
+				continue
+			}
+
+			activeDelegationsAtHeight = append(activeDelegationsAtHeight, gatewayAddress)
+			// Mark the gateway address as added to avoid duplicates.
+			addedDelegations[gatewayAddress] = true
+		}
+
+	}
+
+	// add app address twice to make the ring size of minimum 2
```

Comment:
#PUC. We only add the app address once here, so it is used twice in total when generating the ring. However, saying twice here makes it misleading.

Reply:
Took this case out of the `GetRingAddressesAtBlock` since this function deals with delegatee gateways addresses.
Also made the first addition of the appAddress closer to the second (and to the comment)